dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/twofa/TwoFactorAuthService.java

```diff
@@ -0,0 +1,347 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import javax.annotation.Nonnull;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.common.CodeGenerator;
+import org.hisp.dhis.common.NonTransactional;
+import org.hisp.dhis.email.EmailResponse;
+import org.hisp.dhis.feedback.ConflictException;
+import org.hisp.dhis.feedback.ErrorCode;
+import org.hisp.dhis.feedback.ErrorReport;
+import org.hisp.dhis.feedback.ForbiddenException;
+import org.hisp.dhis.feedback.NotFoundException;
+import org.hisp.dhis.i18n.I18n;
+import org.hisp.dhis.i18n.I18nManager;
+import org.hisp.dhis.message.MessageSender;
+import org.hisp.dhis.outboundmessage.OutboundMessageResponse;
+import org.hisp.dhis.setting.SystemSettingsProvider;
+import org.hisp.dhis.system.velocity.VelocityManager;
+import org.hisp.dhis.user.CurrentUserUtil;
+import org.hisp.dhis.user.SystemUser;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserDetails;
+import org.hisp.dhis.user.UserService;
+import org.hisp.dhis.user.UserSettingsService;
+import org.jboss.aerogear.security.otp.api.Base32;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Morten Svanæs
+ */
+@Slf4j
+@Service
+@AllArgsConstructor
+public class TwoFactorAuthService {
+
+  public static final String TWO_FACTOR_AUTH_REQUIRED_RESTRICTION_NAME = "R_ENABLE_2FA";
+  public static final long TWOFA_EMAIL_CODE_EXPIRY_MILLIS = 900_000; // 15 minutes
+
+  private final SystemSettingsProvider settingsProvider;
+  private final UserService userService;
+  private final MessageSender emailMessageSender;
+  private final UserSettingsService userSettingsService;
+  private final I18nManager i18nManager;
+
+  /**
+   * Enroll user in time-based one-time password (TOTP) 2FA authentication.
+   *
+   * @param username The user that is being enrolled.
+   */
+  @Transactional
+  public void enrollTOTP2FA(@Nonnull String username) throws ConflictException {
+    User user = userService.getUserByUsername(username);
+    if (user == null) {
+      throw new ConflictException(ErrorCode.E6201);
+    }
+    if (user.isTwoFactorEnabled()) {
+      throw new ConflictException(ErrorCode.E3022);
+    }
+    if (!settingsProvider.getCurrentSettings().getTOTP2FAEnabled()) {
+      throw new ConflictException(ErrorCode.E3046);
+    }
+    String totpSeed = Base32.encode(generateSecureRandomBytes(20));
+    user.setSecret(totpSeed);
+    user.setTwoFactorType(TwoFactorType.ENROLLING_TOTP);
+    userService.updateUser(user);
+  }
+
+  /**
+   * Enroll user in email-based 2FA authentication.
+   *
+   * @param username The user that is being enrolled.
+   */
+  @Transactional
+  public void enrollEmail2FA(@Nonnull String username) throws ConflictException {
+    User user = userService.getUserByUsername(username);
+    if (user == null) {
```

Comment:
Not a big deal or necessarily need to change. I have a preference for splitting validation logic and business logic into separate methods. You could have a method `ErrorCode validateEnrollEmail2Fa(..)` that returns null if validation is success and then an `ErrorCode` if it failed. Then potentially throw the exception. That makes it easier to unit test the validation in isolation. Applies further down as well.